### PR TITLE
Migrate Bento components to V1

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #30283.

The main general changes are in the `src/preact/base-element.js` and they are simple:
1. `layoutCallback` is gone.
2. `ensureLoaded` is added. Works via `loading=eager`
3. Loading is started immediately when the Preact component is rendered the first time.
4. The ready state is propagated from the Preact component to AMP element using `Ref<Api>` and `onReadyState` event.

Other changes include:
1. `unloadOnPause` is changed to allow reloading.
2. `VideoWrapper` and `Instagram` components have been updated to V1 protocol.

### CODE REVIEW

File: extensions/amp-instagram/1.0/component.js [hunk 1]
```diff
@@ -15,34 +15,73 @@
...
+      get readyState() {
+        return loaded ? ReadyState.COMPLETE : ReadyState.LOADING;
+      },
+    }),
+    [loaded]
+  );
+
+  // Reset readyState to "laoding" when an iframe is unloaded.
+  const onReadyStateRef = useValueRef(onReadyState);
+  useLayoutEffect(() => {
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-instagram/1.0/component.js:75]: You're right. It'd be more correct here.

File: extensions/amp-instagram/1.0/component.js [hunk 2]
```diff
@@ -15,34 +15,73 @@
...
+    if (!load) {
+      setLoaded(false);
+      const onReadyState = onReadyStateRef.current;
+      if (onReadyState) {
+        onReadyState(ReadyState.LOADING);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-instagram/1.0/component.js:82]: Right. This is why `useLayoutEffect` was important. Basically the risk here is a race condition. We may unload an iframe and schedule the effect for it. But then it could also get reloaded and it'd send its "load" event. The race condition could be between `useEffect(() => setReadyState("loading"))` and `iframe.onload = () => setReadtState("complete")`. The `useLayoutEffect` prevents the race condition.

On the other hand, we can't have an affect that modify the state that it depends on - that might cause the infinite loop.

However, as a compromise I refactored this a bit and moved all `loaded` and event handling into `setLoaded` callback. PTAL and let me know if that's better.

File: extensions/amp-video/1.0/video-wrapper.js [hunk 3]
```diff
@@ -90,14 +92,15 @@ function VideoWrapperWithRef(
     src,
     sources,
     poster,
-    onLoad,
+    onReadyState,
     ...rest
   },
   ref
 ) {
   useResourcesNotify();
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:100]: Still necessary, but strictly for integration with non-V1 components. Hopefully gone soon.

File: extensions/amp-video/1.0/video-wrapper.js [hunk 4]
```diff
@@ -111,6 +114,20 @@ function VideoWrapperWithRef(
   // <source>s change.
   const readyDeferred = useMemo(() => new Deferred(), []);
 
+  const [readyState, setReadyState_] = useState(ReadyState.LOADING);
+
+  const onReadyStateRef = useValueRef(onReadyState);
+  const setReadyState = useCallback(
+    (state, opt_failure) => {
+      setReadyState_(state);
+      const onReadyState = onReadyStateRef.current;
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:123]: Interesting. I'm kind of the opposite :) But in this case it really means the same thing - just has to be propagated slightly weirdly.

File: extensions/amp-video/1.0/video-wrapper.js [hunk 5]
```diff
@@ -111,6 +114,20 @@ function VideoWrapperWithRef(
   // <source>s change.
   const readyDeferred = useMemo(() => new Deferred(), []);
 
+  const [readyState, setReadyState_] = useState(ReadyState.LOADING);
+
+  const onReadyStateRef = useValueRef(onReadyState);
+  const setReadyState = useCallback(
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:120]: It's not very idiomatic, but it happens. The only reason it's done here is to avoid re-firing effects that depend on `setReadyState` because those would indicate "logical" state changes and changing the event handler doesn't constitute a state change. Especially because inlining events is super common in React. E.g. `<Comp onReadyState={() => ...}>`.

File: extensions/amp-video/1.0/video-wrapper.js [hunk 6]
```diff
@@ -130,6 +147,15 @@ function VideoWrapperWithRef(
     setHasUserInteracted(true);
   }, []);
 
+  // Update the initial readyState.
+  useLayoutEffect(() => {
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:151]: Same story: avoiding race conditions with future events. I added a comment. `useLayoutEffect` makes such race conditions impossible.

File: extensions/amp-video/1.0/video-wrapper.js [hunk 7]
```diff
@@ -111,6 +114,20 @@ function VideoWrapperWithRef(
   // <source>s change.
   const readyDeferred = useMemo(() => new Deferred(), []);
 
+  const [readyState, setReadyState_] = useState(ReadyState.LOADING);
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:117]: This is a very good point. I changed it to ref and the same in instagram component.

File: src/custom-element.js [hunk 8]
```diff
@@ -657,7 +660,7 @@ function createBaseCustomElementClass(win) {
           this.signals_.signal(CommonSignals.LOAD_END);
           this.classList.add('i-amphtml-layout');
           this.toggleLoading(false);
-          dom.dispatchCustomEvent(this, 'load');
+          dom.dispatchCustomEvent(this, 'load', null, NO_BUBBLES);
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply src/custom-element.js:663]: It's a public API for Bento. They mimic image/iframe/etc with `element.readyState` and `element.onload`.

File: src/preact/base-element.js [hunk 9]
```diff
@@ -329,51 +336,46 @@ export class PreactBaseElement extends AMP.BaseElement {
...
-  layoutCallback() {
+  ensureLoaded() {
     const Ctor = this.constructor;
     if (!Ctor['loadable']) {
-      return super.layoutCallback();
+      return;
     }
-
     this.mutateProps(dict({'loading': Loading.EAGER}));
+    this.resetLoading_ = true;
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply src/preact/base-element.js:355]: That's a good question and I've been wrestling with this a bit. Here's what happens: `ensureLoaded()` is called and it sets `loading ="eager"`. However, after that it stays "eager" and I thought that's illogical. Once the rendering happen with "eager", IMHO it should go back to "auto" because this is not a user-set state. Also it works together with `unlayoutOnPause`, which I really still hate but don't see a way around it. But the concept is this: we have to kill an iframe to pause, but we still need to be able to reload it when someone asks eager loading.

File: extensions/amp-video/1.0/video-wrapper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (samouri) [reply extensions/amp-video/1.0/video-wrapper.js:100]: When  can it be removed?

File: extensions/amp-video/1.0/video-wrapper.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (samouri) [reply extensions/amp-video/1.0/video-wrapper.js:121]: Which effect would be refired when `onReadyState` is modified? All I see is that the `setReadyState` callback would be redefined.

File: src/preact/base-element.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (samouri) [reply src/preact/base-element.js:355]: What if the user actually does set `loading=eager` (or loading="XYZ") in the interim? Now we've clobbered their value.

File: extensions/amp-video/1.0/video-wrapper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:100]: My expectation that it's actually now only used for the accordion when the element's size can change significantly from render to render. I can't think of any other dependency we have now. And I'd say it'd have to stay until either we move every component to V1 or we try to force through the `intersect-resources` or some alternative to it.

File: extensions/amp-video/1.0/video-wrapper.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:121]: [This one](https://github.com/ampproject/amphtml/pull/32974/files#diff-0f225f57ef9d0e91c443646021b7b37948b85772c99a6b9a74b90223d1c2a650R159)

```
  useLayoutEffect(() => {
    const player = playerRef.current;
    const readyState = player && player.readyState;
    if (readyState != null) {
      setReadyState(readyState > 0 ? ReadyState.COMPLETE : ReadyState.LOADING);
    }
  }, [setReadyState]);
```

We can get into a weird loop here. Say the component is called with inlined `onReadyState` callback like this: `<Comp onReadyState={() => ...}>`. And let's take a case where `onReadyState` can cause re-render. What we get then here is this sequence:

1. The parent component renders `<Comp onRenderState={() => }>`.
2. The `setRenderState` callback is rebuilt.
3. `useLayoutEffect(() => {}, [setRenderState])` is triggered.
4. The parent's `onRenderState` callback is called. It causes the re-render.
5. The parent re-renders `<Comp>` with new `onReadyState`.
6. Repeat.

File: src/preact/base-element.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dvoytenko) [reply src/preact/base-element.js:355]: We currently do not allow setting `loading` value. But when we do then "eager" would have to reset to the value specified by the user instead of "auto", I'd think. In other words this "eager" value is transient when we do `ensureLoaded()`. However, the API doesn't exist yet, so I'm not sure yet how that will change.

File: src/preact/context.js [hunk 10]
```diff
@@ -93,29 +93,24 @@ export function useAmpContext() {
...
-      loading == Loading.EAGER
-      ? true
-      : // Auto: allowed to load.
-      loading == Loading.AUTO && renderable
-      ? true
-      : // Lazy: can continue loading if already started, but do not start it.
-        loadingRef.current || false;
-  loadingRef.current = load;
-  return load;
+  return triggerUnloadOnPauseRef.current ? Loading.UNLOAD : loading;
```
Comment (jridgewell) [inline src/preact/context.js:115]: Do we need a `triggerUnloadOnPauseRef` or `shouldUnloadOnPauseRef`? I think the entire function is reducible to:

```js
  const {loading: loadingContext, playable} = useAmpContext();
  const loading = loadingReducer(loadingProp, loadingContext);
  return loading == Loading.EAGER || playable || !unloadOnPause ? loading : Loading.UNLOAD;
```

File: src/preact/base-element.js [hunk 11]
```diff
@@ -142,34 +143,46 @@ const childIdGenerator = sequentialIdGenerator();
...
   static requiresShadowDom() {
     // eslint-disable-next-line local/no-static-this
     return usesShadowDom(this);
   }
 
+  /** @override @nocollapse */
+  static prerenderAllowed() {
+    // eslint-disable-next-line local/no-static-this
+    const Ctor = this;
+    return !Ctor['loadable'];
```
Comment (jridgewell) [inline src/preact/base-element.js:161]: Remind me please, the `layoutCallback` won't be called until `DisplayObserver` says it's displayable?

File: src/preact/base-element.js [hunk 12]
```diff
@@ -142,34 +143,46 @@ const childIdGenerator = sequentialIdGenerator();
...
+  /** @override @nocollapse */
   static requiresShadowDom() {
     // eslint-disable-next-line local/no-static-this
     return usesShadowDom(this);
   }
 
+  /** @override @nocollapse */
+  static prerenderAllowed() {
+    // eslint-disable-next-line local/no-static-this
+    const Ctor = this;
```
Comment (jridgewell) [inline src/preact/base-element.js:160]: I believe `this` is going to be undefined in production, which is why we added this lint rule. We may need to adjust `prerenderAllowed` to also pass the BaseElement ctor as a param?

File: src/preact/base-element.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jridgewell) [reply src/preact/base-element.js:361]: That makes me think `unlayoutOnPause` is more important than `eager`. When `!playable && unlayoutOnPause`, then `loading` should reset. Otherwise, setting `eager` should be persistent?

File: extensions/amp-video/1.0/video-wrapper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (samouri) [reply extensions/amp-video/1.0/video-wrapper.js:101]: Thinking out loud: would it make sense to hook up a ResOb to `mutateElement` ? Then only `<amp-accordion>` would be calling mutateElement, and only when size changes, instead of each component on each render.

[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:101]: Yeah. That could work, but I need to do a better analysis what we've being using it for. Maybe @jridgewell remembers better. It was always meant to be a temporary thing. Definitely outside of this PR though.

File: src/preact/base-element.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dvoytenko) [reply src/preact/base-element.js:361]: Ok. So there are two things involved: `loading=eager` and `unloadOnPause`. I think we should treat them separately in this thread. This one is specific to the `ensureLoaded()` API which is meant to be transient: once the element is loaded it should no longer require eager loading for the future. Does this make sense? I think `unloadOnPause` is for now a bit confusing and we'll likely be able to improve on it later. But the transient nature of the `ensureLoaded` should be a lot clearer. LMK WYT.

File: extensions/amp-instagram/1.0/component.js [hunk 13]
```diff
@@ -15,34 +15,86 @@
...
+    loading: loadingProp,
+    onReadyState,
+    ...rest
+  },
+  ref
+) {
+  const loading = useLoading(loadingProp, /* unlayoutOnPause */ true);
+  const load = loading !== Loading.UNLOAD;
+
+  const loadedRef = useRef(false);
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-instagram/1.0/component.js:57]: I actually started with the state and changed to the ref based on Jake's recommendation. But after the change, I did like it a bit better. The `value !== loadedRef.current` ended up a minor thing, but the API was inherently event-based and the main goal was to propagate `onReadyState` event, so it ended up more natural. But we can change it back to `useState` + `useEffect`.

File: extensions/amp-instagram/1.0/component.js [hunk 14]
```diff
@@ -15,34 +15,86 @@
...
+  // of `onReadyState` re-triggering the side effects.
+  const onReadyStateRef = useValueRef(onReadyState);
+  const setLoaded = useCallback(
+    (value) => {
+      if (value !== loadedRef.current) {
+        loadedRef.current = value;
+        const onReadyState = onReadyStateRef.current;
+        if (onReadyState) {
+          onReadyState(value ? ReadyState.COMPLETE : ReadyState.LOADING);
+        }
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-instagram/1.0/component.js:68]: awesome!

File: extensions/amp-instagram/1.0/component.js [hunk 15]
```diff
@@ -15,34 +15,86 @@
...
+      // Standard Bento
+      get readyState() {
+        return loadedRef.current ? ReadyState.COMPLETE : ReadyState.LOADING;
+      },
+    }),
+    []
+  );
+
+  // Reset readyState to "loading" when an iframe is unloaded. Has to be
+  // a `useLayoutEffect` to avoid race condition with a future "load" event.
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-instagram/1.0/component.js:91]: I added more details in comments. But basically here's what can happen:

1. An iframe gets unloaded. We schedule `useEffect` to update the state. We don't know when the `useEffect` will actually be called exactly.
2. Then iframe gets reloaded again. Iframe is cached and quickly dispatches the `load` event.
3. The `useEffect` is executed still thinking that the iframe has been unloaded.

So, the race condition is possible between the scheduling for the `useEffect` and the "load" event.

File: extensions/amp-video/1.0/video-iframe.js [hunk 16]
```diff
@@ -97,10 +98,20 @@ function VideoIframeWithRef(
   useImperativeHandle(
     ref,
     () => ({
+      get currentTime() {
+        const playerState = playerStateRef && playerStateRef.current;
+        const value = playerState && playerState['currentTime'];
+        return typeof value == 'number' ? value : NaN;
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-iframe.js:104]: Nice! Time to get used to this :)

File: extensions/amp-video/1.0/video-wrapper.js [hunk 17]
```diff
@@ -90,17 +92,19 @@ function VideoWrapperWithRef(
...
 ) {
   useResourcesNotify();
   const {playable} = useAmpContext();
-  const load = useLoad(loading, unloadOnPause);
+  const loading = useLoading(loadingProp, unloadOnPause);
+  const load = loading !== Loading.UNLOAD;
 
   const [muted, setMuted] = useState(autoplay);
-  const [playing, setPlaying] = useState(false);
+  const [playing, setPlaying_] = useState(false);
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-video/1.0/video-wrapper.js:107]: To make it clear that there's a better way to set this state via the method bellow.

File: extensions/amp-youtube/1.0/component.js [hunk 18]
```diff
@@ -126,20 +126,11 @@ function YoutubeWithRef(
...
-        return info['currentTime'];
-      },
-      get duration() {
-        return info['duration'];
-      },
-    }),
-    [info]
-  );
+  // Player state. Includes `currentTime` and `duration`.
+  const playerStateRef = useRef();
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-youtube/1.0/component.js:130]: Not really super convenient here. The initial code used a `useState`, but really used it more as a ref by muting the value directly. So I decided to make it a ref.

File: extensions/amp-youtube/1.0/component.js [hunk 19]
```diff
@@ -35,7 +35,7 @@ const PlayerStates = {
   '-1': 'unstarted',
   '0': 'ended',
   '1': 'playing',
-  '2': 'paused',
+  '2': 'pause',
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply extensions/amp-youtube/1.0/component.js:38]: I agree. But these are events. Or at least used as events. And the event's name is [pause](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/pause_event), which is also inconsistent in Web spec, since the next event's name is [ended](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/ended_event). But the goal here is to make other video players to dispatch `video`'s events. /cc @alanorozco

File: src/service/resources-impl.js [hunk 20]
```diff
@@ -1199,7 +1199,7 @@ export class ResourcesImpl {
       for (let i = 0; i < this.resources_.length; i++) {
```
[the remark this answers is not in the document]
Comment (dvoytenko) [reply src/service/resources-impl.js:1199]: We are. Just not in this pull request.

File: src/preact/base-element.js (hunk 11 above)
Comment (dvoytenko) [reply src/preact/base-element.js:161]: The `layoutCallback` is not called in V1 at all since it's just a part of a single "build" phase. However, an element can (a) either provide a different `static prerenderAllowed() {}`, or (b) delay some parts of it itself via some observer. Here, I just thought it was a good default - if an element is known not to load anything at all, it's probably safe to prerender it.

File: src/preact/base-element.js (hunk 12 above)
Comment (dvoytenko) [reply src/preact/base-element.js:160]: Can you explain why though? In this case the `prerenderAllowed` is not collapsed or inline or optimized in any way. As the result it's called by caller in PROD as this: `ImplClass.prerenderAllowed()`. That will always have `this` correctly defined, no? I think the issues we have with `static` and `this` is when a static method is optimized as a top-level function.

File: src/preact/context.js (hunk 10 above)
Comment (dvoytenko) [reply src/preact/context.js]: This has been simplified dramatically. It's now even simpler than your suggestion. PTAL.

File: src/preact/base-element.js (hunk 12 above)
Comment (dvoytenko) [reply src/preact/base-element.js:160]: Here's what I see in binaries:

`amp-video-1.0.js`:

```
V.prerenderAllowed=function(){return!this.loadable};
```

And it's called like this in `v0.js`:

```
b.prototype.prerenderAllowed=function(){return ... this.O?this.O.prerenderAllowed(this)};
```

The `this.O` stands for `this.implClass_`.

So from what I can tell - this all looks good. No?

Comment (samouri) [reply src/preact/base-element.js:208]: I think you're right about the issue only being for optimized toplevel funcs

File: extensions/amp-instagram/1.0/component.js [hunk 21]
```diff
@@ -15,34 +15,99 @@
  */
 
 import * as Preact from '../../../src/preact';
```
Comment (samouri) [inline extensions/amp-instagram/1.0/component.js:17]: Why are we porting over amp-instagram instead of just bento components? Can we separate out this diff? If a revert is needed, would hate to have to undo Bento migration as well

File: src/preact/base-element.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (samouri) [reply src/preact/base-element.js:411]: Ah. Is this only relevant in AMP/Bento Modes and not in Preact mode? In which case LGTM

File: src/preact/base-element.js [hunk 22]
```diff
@@ -190,34 +191,46 @@ const IS_EMPTY_TEXT_NODE = (node) =>
...
   static requiresShadowDom() {
     // eslint-disable-next-line local/no-static-this
     return this['usesShadowDom'];
   }
 
+  /** @override @nocollapse */
+  static prerenderAllowed() {
+    // eslint-disable-next-line local/no-static-this
+    const Ctor = this;
+    return !Ctor['loadable'];
```
Comment (jridgewell) [inline src/preact/base-element.js:209]: Shouldn't this be `!!`?
```suggestion
    return !!Ctor['loadable'];
```

File: src/preact/context.js [hunk 23]
```diff
@@ -90,27 +90,9 @@ export function useAmpContext() {
  * Whether the calling component should currently be in the loaded state.
  *
  * @param {!Loading|string} loadingProp
- * @param {boolean} unloadOnPause
  * @return {boolean}
  */
-export function useLoading(loadingProp, unloadOnPause = false) {
```
Comment (jridgewell) [inline src/preact/context.js]: So the fix here is to have the callers take care of `unloadOnPause` themselves? We don't override the loading prop anymore.

File: src/preact/base-element.js [hunk 24]
```diff
@@ -800,6 +806,48 @@ export class PreactBaseElement extends AMP.BaseElement {
...
+    const Ctor = this.constructor;
+    if (Ctor['unloadOnPause']) {
+      this.mutateProps(dict({'loading': Loading.UNLOAD}));
+      this.resetLoading_ = true;
+    } else {
+      const {currentRef_: api} = this;
+      const apiPause = api && api['pause'];
+      if (apiPause) {
+        apiPause();
+      }
```
Comment (jridgewell) [inline src/preact/base-element.js:821]: Nit:
```suggestion
      api?.['pause']?.();
```

File: src/preact/base-element.js [hunk 25]
```diff
@@ -523,23 +520,32 @@ export class PreactBaseElement extends AMP.BaseElement {
...
-      this.loadDeferred_.reject(opt_reason || new Error('load error'));
-      this.loadDeferred_ = null;
-      dispatchCustomEvent(this.element, 'error', null, {bubbles: false});
+  onReadyState_(state, opt_failure) {
+    this.setReadyState(state, opt_failure);
+
+    const Ctor = this.constructor;
+    if (Ctor['unloadOnPause']) {
+      this.updateIsPlaying_(state == ReadyState.COMPLETE);
+    }
```
Comment (jridgewell) [inline src/preact/base-element.js:543]: I don't understand this. What does `unloadOnPause` have to do with the ready state? Did you mean `loadable`?

File: src/preact/base-element.js [hunk 26]
```diff
@@ -523,23 +520,32 @@ export class PreactBaseElement extends AMP.BaseElement {
...
   /** @private */
-  onLoad_() {
-    if (this.loadDeferred_) {
-      this.loadDeferred_.resolve();
-      this.loadDeferred_ = null;
-      dispatchCustomEvent(this.element, 'load', null, {bubbles: false});
+  maybeUpdateReadyState_() {
+    const {currentRef_: api} = this;
+
+    const apiReadyState = api && api['readyState'];
```
Comment (jridgewell) [inline src/preact/base-element.js:526]:
```suggestion
    const apiReadyState = api?.['readyState'];
```

File: extensions/amp-instagram/1.0/component.js [hunk 27]
```diff
@@ -15,34 +15,99 @@
...
+
+  // Reset readyState to "loading" when an iframe is unloaded. Has to be
+  // a `useLayoutEffect` to avoid race condition with a future "load" event.
+  // A race condition can happen when a `useEffect` would be executed
+  // after a future "load" is dispatched.
+  useLayoutEffect(() => {
+    if (!load) {
+      setLoaded(false);
+    }
+  }, [load, setLoaded]);
```
Comment (jridgewell) [inline extensions/amp-instagram/1.0/component.js:98]: This loaded code is really difficult for me to grok. I think what's happening is we want to call `onReadyStateChange` when unloaded->loaded and loaded->unloaded. I don't understand why it can't simply be written like:

```js
const lastLoadedRef = useRef(false);
const [loaded, setLoaded] = useState(false);

if (loaded && !load) {
  onReadyState?.(ReadyState.LOADING);
  lastLoadedRef.current = false;
  setLoaded(false);
} else if (loaded !== lastLoadedRef.current) {
  onReadyState?.(loaded ? ReadyState.COMPLETE : ReadyState.LOADING);
  lastLoadedRef.current = loaded;
}

File: src/preact/base-element.js (hunk 22 above)
Comment (samouri) [reply src/preact/base-element.js:209]: prerender should only be allowed if it _isn't_ loadable. aka it doesn't make any network requests.

File: src/preact/base-element.js (hunk 11 above)
Comment (dvoytenko) [reply src/preact/base-element.js:209]: It's reverse: if there's loadable then preloading will definitely not happen. I just thought it was a good default. If there's nothing to load - then prerendering seems like would be safe in most of cases. But if you think it's too risky - we can remove and let components add it explicitly. LMK.

File: src/preact/base-element.js (hunk 22 above)
Comment (dvoytenko) [reply src/preact/base-element.js:209]: Correct. Only `!loadable` are prerenderable by default. Also, as mentioned in [this thread](https://github.com/ampproject/amphtml/pull/32974#discussion_r588830001), I thought it was a good default. But LMK if you disagree and this is too risky.

File: src/preact/context.js (hunk 23 above)
Comment (dvoytenko) [reply src/preact/context.js]: Correct. That lifecycle is simplified. In many cases the pausing action could be simpler, as I've done with Instagram.

File: src/preact/base-element.js (hunk 24 above)
Comment (dvoytenko) [reply src/preact/base-element.js:821]: This is almost magical :)

File: src/preact/base-element.js (hunk 25 above)
Comment (dvoytenko) [reply src/preact/base-element.js:543]: There are two cases here: 

1. `!onloadOnPause`: a component knows when it's playing exactly and can pause itself. A simplest one - `<video>` tag. In this case we only need track its "displayability" while the component is playing.
2. `unloadOnPause`: an iframe embed. Once it's loaded - we unfortunately not sure if/when it's playing and how to pause it. What this code does: as soon as the component has loaded (`readyState == COMPLETE`) I consider it's playing (or rather "could be playing"). This in turn monitor's the components displayability to pause it when it's no longer displayable (no longer has size).

In other words, this code basically says "we believe a component might be playing at any time once it's loaded".

File: extensions/amp-instagram/1.0/component.js (hunk 27 above)
Comment (dvoytenko) [reply extensions/amp-instagram/1.0/component.js:98]: I renamed things a bit here to make it simpler. LMK if that helps. I renamed `load` variable to `mount` - it basically decides whether or not to actually render/mount the iframe.

The goal of this code specifically is to reset `loaded` to `false` because if the iframe is no longer in DOM, it's not loaded either. And when it's added back, it will restart the whole chain of states again.

Your code might be close, but (a) we can't send events from the component's render function and we don't want to send them on each render. So an effect is better for this.

File: src/preact/base-element.js (hunk 25 above)
Comment (jridgewell) [reply src/preact/base-element.js:543]: Please add that as a comment, that's a good explanation.

**Review by jridgewell — APPROVED** [review]

**Comment by samouri** [comment]
Congrats on merging this!
